model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/HotRodMapStorage.java

```diff
@@ -152,29 +194,91 @@ private static String toOrderString(QueryParameters.OrderBy<?> orderBy) {
 
     @Override
     public Stream<V> read(QueryParameters<M> queryParameters) {
+        if (LockObjectsForModification.isEnabled(session, storedEntityDescriptor.getModelTypeClass())) {
+            return pessimisticQueryRead(queryParameters);
+        }
+
+        Query<E> query = prepareQueryWithPrefixAndParameters(null, queryParameters);
+        CloseableIterator<E> iterator = paginateQuery(query, queryParameters.getOffset(),
+                queryParameters.getLimit()).iterator();
+        return closing(StreamSupport.stream(Spliterators.spliteratorUnknownSize(iterator, 0), false))
+                .onClose(iterator::close)
+                .filter(Objects::nonNull) // see https://github.com/keycloak/keycloak/issues/9271
+                .map(this.delegateProducer);
+    }
+
+    private Stream<V> pessimisticQueryRead(QueryParameters<M> queryParameters) {
+        DefaultModelCriteria<M> dmc = queryParameters.getModelCriteriaBuilder();
+        List<String> idsToRead = new LinkedList<>();
+
+        dmc.partiallyEvaluate((field, operator, operatorArguments) -> {
+            if (MapFieldPredicates.isPrimaryKey(field)) {
+                if (operator == ModelCriteriaBuilder.Operator.EQ) {
+                    idsToRead.add((String) operatorArguments[0]);
+                }
+
+                if (operator == ModelCriteriaBuilder.Operator.IN) {
+                    for (Object operatorArgument : operatorArguments) {
+                        idsToRead.add((String) operatorArgument);
+                    }
+                }
+            }
+
+            return null;
+        });
```

Comment:
The expectation that `idsToRead` will contain IDs which this condition is restricted to is based on premise that any operator on the ID is in positive path and is in the end included in the query. This assumption is not correct though: consider `NOT(ID=1)` condition.

This assumption works in special cases, which should cover the usecases needed here: Assign `true` to every non-ID criteria and let the criteria tree `optimize`. Then pick the IDs from the tree while ensuring no unexpected operation is seen; otherwise defer to the ordinary query. See [`getSingleRestrictionArgument`](https://github.com/keycloak/keycloak/blob/3be2775f9e72845a1f4c178d8c197cc17488b7aa/model/map/src/main/java/org/keycloak/models/map/storage/criteria/DescriptiveModelCriteria.java#L181)